### ISSUE TITLE
Unable to match Protocol implementation with function-level generic to Protocol interface using class-level generic depending on order of function arguments.

### ISSUE THREAD

**Issue by EthanBullVulpe** [microsoft/pyright#6461]
**Describe the bug**
This is related to #5027 , which was fixed. However I have noticed that there are still some situation in which function-level generics in Protocol implementations are not matched to class-level generics in Protocol interfaces. Particularly, the order of a function's arguments can affect if the Protocol implementation is matched correctly.

**Code or Screenshots**
Consider the code below:
```
from abc import abstractmethod
from typing import Protocol, TypeVar

Output = TypeVar("Output")

class ExampleProtocol(Protocol[Output]):

    @abstractmethod
    def method(self, other: Output) -> Output:
        raise NotImplementedError


T = TypeVar("T")
        
class ExampleImplementation:
    
    def method(self, other: T) -> T:
        ...

   
def example_function1(example: ExampleProtocol[Output], other: Output) -> Output:
    ...

def example_function2(other: Output, example: ExampleProtocol[Output]) -> Output:
    ...
    
example_function1(ExampleImplementation(), "abc") # reportGeneralTypeIssues reported here!
example_function2("abc", ExampleImplementation()) # nothing reported here!
```

Both `example_function1` and `example_function2` have essentially the same signature - except that the order of arguments is switched. Pyright reports a "reportGeneralTypeIssues" for the call to `example_function1`, with the message:
```
Argument of type "ExampleImplementation" cannot be assigned to parameter "example" of type "ExampleProtocol[Output@example_function1]" in function "example_function1"
  "ExampleImplementation" is incompatible with protocol "ExampleProtocol[T@method | str]"
    Type parameter "Output@ExampleProtocol" is invariant, but "T@method" is not the same as "T@method | str"
      Pylance(reportGeneralTypeIssues )
```
Pyright accepts the call to `example_function2`.

In my opinion, both the function calls should be accepted, since clearly the type of `Output` in the call to example_function1 is `str`, and `ExampleImplementation` clearly implements `ExampleProtocol[str]`. However, some funny business happens, and pyright seems to think the type of `Output` is `T@method | str`, rather than just `str`, which (I think) leads to the above error. If you switch the order of the arguments, as in `example_function2`, then pyright easily recognizes that the type of `Output`  is `str`, and no issue is reported.

**VS Code extension or command-line**
Python version 3.10, pyright (v1.1.335) is being run though the pylance (v2023.11.12) extension in vscode.

Thanks!

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
